### DISCUSSION TITLE
use openid login

### DISCUSSION THREAD

**Discussion by alan7622** [fossology/fossology#1957] [Q&A] [answered]
Is it still available to use openid login fossology?

**Comment by GMishx** [comment]
At the moment FOSSology only supports login from username/password or LDAP: https://github.com/fossology/fossology/wiki/External-authentication-configuration

**Comment by NicolasToussaint** [comment]
Actually, the external authentication in place allows any type of authentication supported by Apache.

Apache is configured to handle the authentication step, and then injects the resulting Headers in the HTTP query. The headers are used by Fossology to log the correct user in.

See general workflow here: https://github.com/fossology/fossology/wiki/External-authentication-configuration#general-workflow

**Reply by NicolasToussaint** [reply]
btw, the above link provides examples for LDAP and SAML/Shibboleth authentication (currently in production use).
I just updated the page to simplify the LDAP configuration example.

**Reply by GMishx** [reply]
Ah, my bad. Thank you Nicolas.

**Reply by alan7622** [reply]
Hello, thank you for noticing I used an apache mod_auth_openidc module.
I integrate it inside the fossology container, and successfully comeout with the login page.
my login page is at: http://localhost:80
my redirectURI is at: http://localhost:80/repo/

After I login, I will redirect to localhost:8082 and not localhost:8082/repo. Are there any settings in the config that may cause this problem? 
And the second question is after I redirect to localhost:8082/repo successfully.  I just need to set the CONF_EXT_AUTH_ENABLE=true, in the path install/defconf/fossology.conf.in and I can login?, Is my flow correct thank you so much?

LoadModule auth_openidc_module /usr/lib/apache2/modules/mod_auth_openidc.so
OIDCProviderMetadataURL XXXXXXXXXXXX
OIDCClientID XXXXXXXXXXXXXXX
OIDCClientSecret XXXXXXXXXXXXX
OIDCRedirectURI http://localhost:8082/repo/
OIDCCryptoPassphrase alan7622222223434
<Location />
    AuthType openid-connect
    require valid-user
</Location>

**Reply by GMishx** [reply]
Reading instructions from https://github.com/zmartzone/mod_auth_openidc#quickstart-with-a-generic-openid-connect-provider

The configuration you done looks okay to me. But by default, localhost:80 does not server FOSSology. It only loads at localhost:80/repo. So your redirect URI should be localhost:80/repo

Then you need to set `CONF_EXT_AUTH_ENABLE=true` and update the `CONF_EXT_AUTH_ENV_USER` to the variable containing the userid for login.

**Comment by alan7622** [comment, marked as answer]
Is there any way I can set Header always set Cache-Control "no-cache, no-store, max-age=0", in the apache folder inside fossology, thank you.

**Reply by GMishx** [reply]
You need to add `Header append Cache-Control "no-cache, no-store, max-age=0"` to Apache configuration inside `<Directory></Directory>`, enable `mod_headers` (`sudo a2enmod headers`) and restart Apache.

**Reply by alan7622** [reply]
@GMishx, thank you for your response. I have two questions. I integrate the apache openid module and the opendid config file inside the fossology container, which means I'm using an internal authentication or still an external authentication ?, thank you so much
1.which url should be my redirect uri 1. http://localhost:8082/repo/api 2. http://localhost:8082/repo/
2. Keep getting error, which means I can't authenticate my user 
 oidc_handle_authorization_response: session created and stored, returning to original URL
[Mon May 10 07:26:01.809434 2021] [authz_core:debug] [pid 1048] mod_authz_core.c(809): [client 10.10.0.1:37772] AH01626: authorization result of Require valid-user : denied (no authenticated user yet), referer: http://localhost:8082/repo/
[Mon May 10 07:26:01.809462 2021] [authz_core:debug] [pid 1048] mod_authz_core.c(809): [client 10.10.0.1:37772] AH01626: authorization result of <RequireAny>: denied (no authenticated user yet), referer: http://localhost:8082/repo/

**Reply by GMishx** [reply]
1. Any authentication other than FOSSology's username & password should be external.
2. You should redirect to localhost:8082/repo/ I think.
3. The missing authentication can be result of misconfigured plugin. Please check your Apache server's error log.

**Reply by alan7622** [reply]
@NicolasToussaint @GMishx Thank you guys so much, finally I found out the reason, it is because the domain of fossology and my openid config file has a conflict and it finally works, thank you so much.

**Reply by alan7622** [reply]
I have a question after using the external authentication login, is there any way to access the Default Administrator access, because it seems that after I login with the external authentication, I don't have the authority to select the other user's access level and also select the other user's top-level folder, thank you a lot

**Reply by GMishx** [reply]
You can try changing the username of the default administrator (fossy) to your email address.
Otherwise, the authentication can temporarily be changed to password and you can assign other as administrators. Then change to external authentication again.

**Reply by alan7622** [reply]
@GMishx I first login with the openid login and disconfig the openid login and login with the original login and set myself to full adminstrator, then I got the full administrator after using the openid login, thank you so much

**Reply by NicolasToussaint** [reply]
@alan7622, you can also hack the database directly :-)

`update users set user_perm = 10 where user_name = 'my-user-name';`

**Reply by alan7622** [reply]
@NicolasToussaint smart way, I use this way to solve the problem thanks

**Comment by alan7622** [comment]
I met about an issue about the openid login with the same code, which works normally before.  It seems that the issue is because I pull the latest image fossology/fossology:latest from Docker hub, but when I use the fossology/fossology:3.10.0 base image it works fine. @GMishx @NicolasToussaint It seems the error is from the php code, do you guys have any idea thank you so much.
[Fri May 28 07:26:28.538244 2021] [php7:notice] [pid 344] [client 10.10.0.1:44870] PHP Notice:  Undefined index:  in /usr/local/share/fossology/lib/php/common-auth.php on line 71
[Fri May 28 07:26:28.617052 2021] [php7:error] [pid 344] [client 10.10.0.1:44870] PHP Fatal error:  Uncaught ArgumentCountError: Too few arguments to function add_user(), 8 passed in /usr/local/share/fossology/www/ui/core-auth.php on line 292 and at least 9 expected in /usr/local/share/fossology/lib/php/common-users.php:43\nStack trace:\n#0 /usr/local/share/fossology/www/ui/core-auth.php(292): add_user('alan.hsieh@...', NULL, '$2y$10$/mABOGoI...', 3, 'Alan.Hsieh@...', 'y', '', 1)\n#1 /usr/local/share/fossology/www/ui/core-auth.php(96): core_auth->checkUsernameAndPassword('alan.hsieh@...', 'd5b59c118d17c6e...')\n#2 /usr/local/share/fossology/lib/php/Plugin/FO_Plugin.php(476): core_auth->PostInitialize()\n#3 /usr/local/share/fossology/lib/php/common-plugin.php(229): FO_Plugin->preInstall()\n#4 /usr/local/share/fossology/www/ui/index.php(57): plugin_preinstall()\n#5 {main}\n  thrown in /usr/local/share/fossology/lib/php/common-users.php on line 43
